### ISSUE TITLE
Nearly 3-hour CI runs

### ISSUE THREAD

**Issue by brenns10** [osandov/drgn#365]
Some of the CI runs have been taking as much as 3 hours. It's not specific to clang or a Python version. It looks like the VMs are spending a lot of time in the logs for computing the loops per jiffies / BogoMIPS thing? I made a script to find large time gaps in the logs, and print what kernel version was last booted along with the context of about 10 lines:

```python
"""
Find long time gaps in Github Actions CI Logs

Go to the CI run, use the gear button at the top right and click "View
raw logs". Save the resulting file to disk and pipe it into this script.

EG: https://github.com/osandov/drgn/actions/runs/6725563123/job/18289675003?pr=364
"""
import datetime
import sys


lines = []
context_left = 0
last_ts = None
current_linux = None

for line in sys.stdin:
    should_print = False

    if context_left > 0:
        should_print = True
        context_left -= 1
    ts, rest = line.split(" ", 1)
    if rest.startswith("Linux version "):
        current_linux = rest.split()[2]
    ts = datetime.datetime.fromisoformat(ts[:-2])
    if last_ts and (ts - last_ts) >= datetime.timedelta(minutes=3):
        should_print = True
        context_left = 10
        if lines:
            print(f"\n\n  LINUX VERSION: {current_linux}\n\n")
            for line in lines:
                print(line, end="")
        lines = []


    if should_print:
        print(line, end="")
    else:
        lines.append(line)
        lines = lines[-10:]

    last_ts = ts
```

Here's some findings from #364 - clang - python 3.11 run:

```
  LINUX VERSION: 5.13.19-vmtest22.1default


2023-11-02T06:53:06.3048792Z Spectre V1 : Mitigation: usercopy/swapgs barriers and __user pointer sanitization
2023-11-02T06:53:06.3051168Z Spectre V2 : Mitigation: Full AMD retpoline
2023-11-02T06:53:06.3053593Z Spectre V2 : Spectre v2 / SpectreRSB mitigation: Filling RSB on context switch
2023-11-02T06:53:06.3054759Z Speculative Store Bypass: Vulnerable
2023-11-02T06:53:06.3151395Z Freeing SMP alternatives memory: 12K
2023-11-02T06:53:06.4316388Z smpboot: CPU0: AMD QEMU Virtual CPU version 2.5+ (family: 0x6, model: 0x6, stepping: 0x3)
2023-11-02T06:53:06.4365861Z Performance Events: PMU not available due to virtualization, using software events only.
2023-11-02T06:53:06.4377137Z rcu: Hierarchical SRCU implementation.
2023-11-02T06:53:06.4400732Z smp: Bringing up secondary CPUs ...
2023-11-02T06:53:06.4421135Z x86: Booting SMP configuration:
2023-11-02T06:53:06.4421135Z x86: Booting SMP configuration:
2023-11-02T07:07:20.2557292Z calibrate_delay_direct() dropping max bogoMips estimate 2 = 244684371590
2023-11-02T07:07:20.2558419Z calibrate_delay_direct() dropping max bogoMips estimate 1 = 244166718848
2023-11-02T07:07:20.2559665Z calibrate_delay_direct() dropping max bogoMips estimate 4 = 47428804157
2023-11-02T07:07:20.2560734Z calibrate_delay_direct() dropping max bogoMips estimate 0 = 127183409
2023-11-02T07:07:20.2561842Z calibrate_delay_direct() failed to get a good estimate for loops_per_jiffy.
2023-11-02T07:07:20.2563174Z Probably due to long platform interrupts. Consider using "lpj=" boot option.
2023-11-02T07:21:33.0533223Z  #2
2023-11-02T07:21:33.0535256Z calibrate_delay_direct() dropping max bogoMips estimate 1 = 144966717195
2023-11-02T07:21:33.0537383Z calibrate_delay_direct() dropping max bogoMips estimate 4 = 111805432154
2023-11-02T07:21:33.0539363Z calibrate_delay_direct() dropping max bogoMips estimate 3 = 147274810
2023-11-02T07:31:48.0049750Z  #3
2023-11-02T07:31:48.0051567Z calibrate_delay_direct() dropping max bogoMips estimate 4 = 130217933417
2023-11-02T07:31:48.0053768Z calibrate_delay_direct() dropping max bogoMips estimate 2 = 42555223884
2023-11-02T07:31:48.0055875Z calibrate_delay_direct() dropping min bogoMips estimate 3 = 63652899
2023-11-02T07:31:48.0062657Z smp: Brought up 1 node, 4 CPUs
2023-11-02T07:31:48.0064136Z smpboot: Max logical packages: 4
2023-11-02T07:31:48.0066832Z smpboot: Total of 4 processors activated (39343786.88 BogoMIPS)
2023-11-02T07:31:48.0185688Z devtmpfs: initialized
2023-11-02T07:31:48.0244890Z clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
2023-11-02T07:31:48.0246839Z futex hash table entries: 1024 (order: 4, 65536 bytes, linear)
2023-11-02T07:31:48.0269989Z NET: Registered protocol family 16


  LINUX VERSION: 5.12.19-vmtest22.1default


2023-11-02T07:32:36.1642382Z Spectre V1 : Mitigation: usercopy/swapgs barriers and __user pointer sanitization
2023-11-02T07:32:36.1644530Z Spectre V2 : Mitigation: Full AMD retpoline
2023-11-02T07:32:36.1646896Z Spectre V2 : Spectre v2 / SpectreRSB mitigation: Filling RSB on context switch
2023-11-02T07:32:36.1648059Z Speculative Store Bypass: Vulnerable
2023-11-02T07:32:36.1680539Z Freeing SMP alternatives memory: 12K
2023-11-02T07:32:36.2856329Z smpboot: CPU0: AMD QEMU Virtual CPU version 2.5+ (family: 0x6, model: 0x6, stepping: 0x3)
2023-11-02T07:32:36.2906062Z Performance Events: PMU not available due to virtualization, using software events only.
2023-11-02T07:32:36.2918021Z rcu: Hierarchical SRCU implementation.
2023-11-02T07:32:36.2941602Z smp: Bringing up secondary CPUs ...
2023-11-02T07:32:36.2960579Z x86: Booting SMP configuration:
2023-11-02T07:32:36.2960579Z x86: Booting SMP configuration:
2023-11-02T07:52:51.6954265Z calibrate_delay_direct() dropping max bogoMips estimate 4 = 494358598477
2023-11-02T07:52:51.6956656Z calibrate_delay_direct() dropping max bogoMips estimate 2 = 254532895149
2023-11-02T07:52:51.6958735Z calibrate_delay_direct() dropping min bogoMips estimate 1 = 24453476
2023-11-02T08:01:11.8311405Z  #2
2023-11-02T08:01:11.8313441Z calibrate_delay_direct() dropping max bogoMips estimate 2 = 326288435405
2023-11-02T08:01:11.8314830Z calibrate_delay_direct() dropping max bogoMips estimate 4 = 288598642
2023-11-02T08:01:11.8316046Z calibrate_delay_direct() dropping max bogoMips estimate 3 = 136967603
2023-11-02T08:09:31.9951195Z  #3
2023-11-02T08:09:31.9952830Z calibrate_delay_direct() dropping max bogoMips estimate 0 = 259303869
2023-11-02T08:09:31.9954870Z calibrate_delay_direct() dropping min bogoMips estimate 4 = 29353939
2023-11-02T08:09:31.9956619Z calibrate_delay_direct() dropping max bogoMips estimate 3 = 185907553
2023-11-02T08:09:31.9960645Z calibrate_delay_direct() dropping max bogoMips estimate 2 = 122304925
2023-11-02T08:09:31.9963277Z calibrate_delay_direct() failed to get a good estimate for loops_per_jiffy.
2023-11-02T08:09:31.9965198Z Probably due to long platform interrupts. Consider using "lpj=" boot option.
2023-11-02T08:18:14.6446480Z smp: Brought up 1 node, 4 CPUs
2023-11-02T08:18:14.6447590Z smpboot: Max logical packages: 4
2023-11-02T08:18:14.6450289Z smpboot: Total of 4 processors activated (29096283.49 BogoMIPS)
2023-11-02T08:18:14.6573138Z devtmpfs: initialized
2023-11-02T08:18:14.6632418Z clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
2023-11-02T08:18:14.6635887Z futex hash table entries: 1024 (order: 4, 65536 bytes, linear)
2023-11-02T08:18:14.6658800Z NET: Registered protocol family 16
2023-11-02T08:18:14.6701865Z thermal_sys: Registered thermal governor 'step_wise'
2023-11-02T08:18:14.6722405Z thermal_sys: Registered thermal governor 'user_space'
2023-11-02T08:18:14.6723100Z cpuidle: using governor ladder
2023-11-02T08:18:14.6730091Z ACPI: bus type PCI registered


  LINUX VERSION: 5.11.22-vmtest22.1default


2023-11-02T08:19:02.2586631Z Spectre V1 : Mitigation: usercopy/swapgs barriers and __user pointer sanitization
2023-11-02T08:19:02.2587983Z Spectre V2 : Mitigation: Full AMD retpoline
2023-11-02T08:19:02.2589887Z Spectre V2 : Spectre v2 / SpectreRSB mitigation: Filling RSB on context switch
2023-11-02T08:19:02.2591025Z Speculative Store Bypass: Vulnerable
2023-11-02T08:19:02.2622825Z Freeing SMP alternatives memory: 12K
2023-11-02T08:19:02.3798058Z smpboot: CPU0: AMD QEMU Virtual CPU version 2.5+ (family: 0x6, model: 0x6, stepping: 0x3)
2023-11-02T08:19:02.3847422Z Performance Events: PMU not available due to virtualization, using software events only.
2023-11-02T08:19:02.3859116Z rcu: Hierarchical SRCU implementation.
2023-11-02T08:19:02.3882840Z smp: Bringing up secondary CPUs ...
2023-11-02T08:19:02.3901985Z x86: Booting SMP configuration:
2023-11-02T08:19:02.3901985Z x86: Booting SMP configuration:
2023-11-02T08:31:21.1488975Z calibrate_delay_direct() dropping max bogoMips estimate 4 = 768540205654
2023-11-02T08:31:21.1492088Z calibrate_delay_direct() dropping max bogoMips estimate 1 = 146739819
2023-11-02T08:31:21.1494017Z calibrate_delay_direct() dropping max bogoMips estimate 2 = 93024835
2023-11-02T08:31:21.1496206Z calibrate_delay_direct() dropping max bogoMips estimate 3 = 78346214
2023-11-02T08:31:21.1498976Z calibrate_delay_direct() failed to get a good estimate for loops_per_jiffy.
2023-11-02T08:31:21.1501260Z Probably due to long platform interrupts. Consider using "lpj=" boot option.
2023-11-02T08:54:09.3871237Z  #2
2023-11-02T08:54:09.3873766Z calibrate_delay_direct() dropping max bogoMips estimate 0 = 1217661456540
2023-11-02T08:54:09.3876484Z calibrate_delay_direct() dropping max bogoMips estimate 4 = 499018906
2023-11-02T08:54:09.3879106Z calibrate_delay_direct() dropping max bogoMips estimate 3 = 229909043
2023-11-02T08:54:09.3881721Z calibrate_delay_direct() dropping max bogoMips estimate 2 = 34237567
2023-11-02T08:54:09.3884493Z calibrate_delay_direct() failed to get a good estimate for loops_per_jiffy.
2023-11-02T08:54:09.3886962Z Probably due to long platform interrupts. Consider using "lpj=" boot option.
2023-11-02T09:10:48.7311387Z  #3
2023-11-02T09:10:48.7313281Z calibrate_delay_direct() dropping max bogoMips estimate 0 = 1179270661384
2023-11-02T09:10:48.7315637Z calibrate_delay_direct() dropping max bogoMips estimate 4 = 254396059
2023-11-02T09:10:48.7318157Z calibrate_delay_direct() dropping min bogoMips estimate 1 = 44022946
2023-11-02T09:10:48.7324668Z smp: Brought up 1 node, 4 CPUs
2023-11-02T09:10:48.7325477Z smpboot: Max logical packages: 4
2023-11-02T09:10:48.7327511Z smpboot: Total of 4 processors activated (584933.88 BogoMIPS)
2023-11-02T09:10:48.7454856Z devtmpfs: initialized
2023-11-02T09:10:48.7514560Z clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
2023-11-02T09:10:48.7517265Z futex hash table entries: 1024 (order: 4, 65536 bytes, linear)
2023-11-02T09:10:48.7539920Z NET: Registered protocol family 16
```

I also looked at the gcc - python 3.6 run (both were slow). It was the same 3 kernels: 5.11, 5.12, 5.13. We may have hit a regression?

**Comment by brenns10** [comment]
The latest vmtest kernels were built [here](https://github.com/osandov/drgn/actions/runs/6689158411):
```
 2023-10-30 06:30:35,499:INFO:__main__:latest kernel versions: v6.6, v6.5.9, v6.4.16, v6.3.13, v6.2.16, v6.1.60, v6.0.19, v5.19.17, v5.18.19, v5.17.15, v5.16.20, v5.15.137, v5.14.21, v5.13.19, v5.12.19, v5.11.22, v5.10.199, v5.4.259, v4.19.297, v4.14.328, v4.9.337
2023-10-30 06:30:35,500:INFO:__main__:kernel versions to build:
2023-10-30 06:30:35,500:INFO:__main__:  v6.6 (x86_64 [default])
2023-10-30 06:30:35,500:INFO:__main__:  v6.5.9 (x86_64 [default])
2023-10-30 06:30:35,500:INFO:__main__:  v6.1.60 (x86_64 [default])
2023-10-30 06:30:35,500:INFO:__main__:  v5.15.137 (x86_64 [default])
2023-10-30 06:30:35,500:INFO:__main__:  v5.10.199 (x86_64 [default])
2023-10-30 06:30:35,500:INFO:__main__:  v5.4.259 (x86_64 [default])
2023-10-30 06:30:35,500:INFO:__main__:  v4.19.297 (x86_64 [default])
2023-10-30 06:30:35,500:INFO:__main__:  v4.14.328 (x86_64 [default])
```

Interestingly, none of those 3 kernels were built. So maybe this isn't a regression in the kernel? In any case, we could probably  find some way to skip this, maybe with the lpj= option?

**Comment by osandov** [comment]
Thanks for looking into this. I've also been getting full-blown timeout failures. Is it always those 3 kernels? I think those have been EOL for awhile, so they haven't been rebuilt since the last config changes a couple of months ago. I wonder if QEMU was updated recently.

**Comment by brenns10** [comment]
I will have to get back to you on whether it's _always_ those 3 kernels, I can check more logs in the morning. But I did figure I could try to reproduce it on my home laptop. I see the following during that part of the boot process:
```
rcu: Hierarchical SRCU implementation.
smp: Bringing up secondary CPUs ...
x86: Booting SMP configuration:
.... node  #0, CPUs:      #1
kvm-clock: cpu 1, msr 37df1041, secondary cpu clock
kvm-guest: KVM setup async PF for cpu 1
kvm-guest: stealtime: cpu 1, msr 7fc5b580
 #2
kvm-clock: cpu 2, msr 37df1081, secondary cpu clock
kvm-guest: KVM setup async PF for cpu 2
kvm-guest: stealtime: cpu 2, msr 7fc9b580
 #3
kvm-clock: cpu 3, msr 37df10c1, secondary cpu clock
kvm-guest: KVM setup async PF for cpu 3
kvm-guest: stealtime: cpu 3, msr 7fcdb580
 #4
kvm-clock: cpu 4, msr 37df1101, secondary cpu clock
kvm-guest: KVM setup async PF for cpu 4
kvm-guest: stealtime: cpu 4, msr 7fd1b580
 #5
kvm-clock: cpu 5, msr 37df1141, secondary cpu clock
kvm-guest: KVM setup async PF for cpu 5
kvm-guest: stealtime: cpu 5, msr 7fd5b580
 #6
kvm-clock: cpu 6, msr 37df1181, secondary cpu clock
kvm-guest: KVM setup async PF for cpu 6
kvm-guest: stealtime: cpu 6, msr 7fd9b580
 #7
kvm-clock: cpu 7, msr 37df11c1, secondary cpu clock
kvm-guest: KVM setup async PF for cpu 7
kvm-guest: stealtime: cpu 7, msr 7fddb580
```

That doesn't match Github actions, probably because they don't have KVM. So I patched vmtest to not use KVM even if it's available:

```
diff --git a/vmtest/vm.py b/vmtest/vm.py
index 917d8345..53e059d6 100644
--- a/vmtest/vm.py
+++ b/vmtest/vm.py
@@ -215,7 +215,8 @@ def run_in_vm(
 
     kvm_args = []
     if HOST_ARCHITECTURE is not None and kernel.arch.name == HOST_ARCHITECTURE.name:
-        if os.access("/dev/kvm", os.R_OK | os.W_OK):
+        #if os.access("/dev/kvm", os.R_OK | os.W_OK):
+        if False:
             kvm_args = ["-cpu", "host", "-enable-kvm"]
         else:
             print(
```

And... yeah, wow, now I immediately get pauses at:
```
rcu: Hierarchical SRCU implementation.
smp: Bringing up secondary CPUs ...
x86: Booting SMP configuration:

```

I've tried it on the following:

- 6.6 - no pause
- 6.5 - no pause
- 6.4 - no pause
- 6.3 - no pause
- 6.2 - no pause
- 6.1 - no pause
- 6.0 - no pause
- 5.19 - no pause
- 5.18 - no pause
- 5.17 - no pause
- 5.16 - no pause
- 5.15 - no pause
- 5.14 - no pause
- 5.13 - pause
- 5.12 - pause
- 5.11 - pause
- 5.10 - no pause
- 5.4 - no pause
- 4.19 - no pause
- 4.14 - no pause
- 4.9 - no pause

Seems like it's definitely something to do with 5.11-5.13, with KVM disabled. That's probably enough evening debugging for me tonight but seems like a good lead.

**Comment by osandov** [comment]
I was able to reproduce this locally as well, both on my laptop and my development server, with QEMU 7.something and QEMU 4.2.0, and with the latest and oldest available vmtest builds of 5.11, 5.12, and 5.13. So this really doesn't seem to be new, although I am not remember seeing it before this week.

Regardless, since it's kernel version-specific, we can hopefully backport a fix to apply in `vmtest.kbuild`.

**Comment by brenns10** [comment]
I figured I could take a QEMU core dump to try to figure out what was on CPU during these hangs. But unfortunately, the vmcoreinfo hasn't even been initialized at this point in the boot process, so Drgn couldn't immediately open it.

I was able to work around this by using my `vmcoreinfo` branch that lets me specify a vmcoreinfo text file, as well as an architecture, to work better with QEMU dumps. I also added `nokaslr` to the command line. I used the following to generate a skeleton vmcoreinfo with makedumpfile:

```
 makedumpfile -g bringup.vi.txt -x build/vmtest/x86_64/kernel-5.11.22-vmtest22.1default/build/vmlinux
```

Then I added the correct value of `swapper_pg_dir` (it doesn't change from boot to boot if you disable kaslr) and I added KERNELOFFSET:

```
SYMBOL(swapper_pg_dir)=ffffffff8200a000
KERNELOFFSET=0
```

Actual generation of the vmcore is done with KVM disabled, and nokaslr added to the cmdline. At the hang, I did Ctrl-A C to bring up the console, then `dump-guest-memory -z bringup.flat` to generate the vmcore, and `q` to exit. Then `makedumpfile -R bringup <bringup.flat` to reassemble it, and finally I could open it with drgn:

```
$ python -m drgn -c bringup -a x86_64 -i bringup.vi.txt -s build/vmtest/x86_64/kernel-5.11.22-vmtest22.1default/build/vmlinux
drgn 0.0.24+70.ge8d4a0ee.dirty (using Python 3.11.5, elfutils 0.189, with libkdumpfile)
For help, type help(drgn).
>>> import drgn
>>> from drgn import NULL, Object, cast, container_of, execscript, offsetof, reinterpret, sizeof
>>> from drgn.helpers.common import *
>>> from drgn.helpers.linux import *
>>> prog.stack_trace(cpu_curr(prog, 0))
#0  __raw_spin_unlock_irq (./include/linux/spinlock_api_smp.h:170:1)
#1  __schedule (./kernel/sched/core.c:5079:3)
#2  schedule (./kernel/sched/core.c:5152:3)
#3  do_boot_cpu (./arch/x86/kernel/smpboot.c:1122:4)
#4  native_cpu_up (./arch/x86/kernel/smpboot.c:1180:8)
#5  __cpu_up (./arch/x86/include/asm/smp.h:87:9)
#6  bringup_cpu (./kernel/cpu.c:561:8)
#7  cpuhp_invoke_callback (./kernel/cpu.c:173:9)
#8  cpuhp_up_callbacks (./kernel/cpu.c:615:9)
#9  _cpu_up (./kernel/cpu.c:1246:8)
#10 cpu_up (./kernel/cpu.c:1281:8)
#11 bringup_nonboot_cpus (./kernel/cpu.c:1343:4)
#12 smp_init (./kernel/smp.c:814:2)
#13 kernel_init_freeable (./init/main.c:1526:2)
#14 kernel_init (./init/main.c:1421:2)
#15 ret_from_fork+0x22/0x2d (./arch/x86/entry/entry_64.S:296)
>>> prog.stack_trace(cpu_curr(prog, 1))
#0  calibrate_delay_direct (./init/calibrate.c:76:4)
#1  calibrate_delay (./init/calibrate.c:297:20)
#2  smp_callin (./arch/x86/kernel/smpboot.c:202:2)
#3  start_secondary (./arch/x86/kernel/smpboot.c:240:2)
#4  secondary_startup_64+0xc7/0xc7 (./arch/x86/kernel/head_64.S:283)
```

So that's what's on CPU for the hang, on the 5.11 kernel at least.
```
```

**Comment by osandov** [comment]
One more interesting fact, setting the number of CPUs to 1 (passing -smp 1 to QEMU) seems to "fix" it. Can you get stack traces from the other CPUs?

I wonder if this is something we could bisect. It can be tricky to build old kernel versions with recent toolchains, though...

**Comment by osandov** [comment]
Oops, I missed that you did include a call trace from CPU 1, scratch that part of my comment (really cool that you were able to recover this, btw).

**Comment by brenns10** [comment]
CPUs 2-7 didn't have anything useful:
```
>>> prog.stack_trace(cpu_curr(prog, 2))
#0  0xfd5b4
>>> prog.stack_trace(cpu_curr(prog, 3))
#0  0xfd5b4
>>> prog.stack_trace(cpu_curr(prog, 4))
#0  0xfd5b4
>>> prog.stack_trace(cpu_curr(prog, 5))
#0  0xfd5b4
>>> prog.stack_trace(cpu_curr(prog, 6))
#0  0xfd5b4
>>> prog.stack_trace(cpu_curr(prog, 7))
#0  0xfd5b4
```
Probably they haven't been brought up yet, looks like it's going in serial?

**Comment by osandov** [comment]
That makes sense. Lol, `init/calibrate.c` hasn't changed since 2014. My guess is that it's something timer related. Do you want to try bisecting, or should I kick something off?

**Comment by brenns10** [comment]
I'm not fully familiar with the vmtest kbuild yet, so it may be faster for you to try.

**Comment by brenns10** [comment]
Ok, it's not actually that hard to do :)

I just verified upstream v5.11 is BAD - so the issue is not on the stable branch. I'll verify v5.4 is GOOD and start bisecting, unless you've already done so.

**Comment by osandov** [comment]
I have `git bisect run` going between v5.13 (hangs) and v5.14 (does not hang).

**Comment by brenns10** [comment]
Haha I guess it's between you finding the fix or me finding the cause :P

**Comment by osandov** [comment]
The fix bisected to torvalds/linux@5cb9eaa3d274f75539077a28cf01e3563195fa53, which makes no sense... But the parent commit, torvalds/linux@39d371b7c0c299d489041884d005aacc4bba8c15, hangs, and that one doesn't hang. I'm so confused.

**Comment by brenns10** [comment]
Ok, the cause bisected to torvalds/linux@2558aacff8586699bcd248b406febb28b0a25de2 as the first bad (hanging) commit. That doesn't help me much either...

**Comment by osandov** [comment]
Yup, torvalds/linux@2558aacff8586699bcd248b406febb28b0a25de2 hangs here, too, and its parent doesn't hang.

To make this even crazier, it doesn't hang when compiled with Clang!!

**Comment by osandov** [comment]
Applying this part of https://github.com/torvalds/linux/commit/5cb9eaa3d274f75539077a28cf01e3563195fa53:

```diff
diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 5568018222d9..5e6f5f5750a3 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -5164,7 +5164,7 @@ static void __sched notrace __schedule(bool preempt)
 
 		rq_unpin_lock(rq, &rf);
 		__balance_callbacks(rq);
-		raw_spin_unlock_irq(&rq->lock);
+		raw_spin_rq_unlock_irq(rq);
 	}
 }
 
```

on top of https://github.com/torvalds/linux/commit/39d371b7c0c299d489041884d005aacc4bba8c15 "fixes" the hang.

**Comment by osandov** [comment]
Comparing the disassembly, without the "fix" we're missing an `sti`:

```
Changed symbols:
__schedule
--- busted.o
+++ fixed.o
@@ -235,7 +235,7 @@
 	call   <__schedule+0x409>
 	mov    %rax,%rdi
 	call   <__schedule+0x411>
-	jmp    <__schedule+0x43e>
+	jmp    <__schedule+0x43f>
 	mov    0x880(%rbx),%eax
 	and    $0xfffffffc,%eax
 	mov    %eax,0x880(%rbx)
@@ -246,10 +246,11 @@
 	call   <__schedule+0x436>
 	mov    %rbx,%rdi
 	call   <__schedule+0x43e>
+	sti    
 	mov    -0x30(%rbp),%rax
 	xor    %gs:0x28,%rax
-	je     <__schedule+0x452>
-	call   <__schedule+0x452>
+	je     <__schedule+0x453>
+	call   <__schedule+0x453>
 	add    $0x20,%rsp
 	pop    %rbx
 	pop    %r12
```

So `raw_spin_unlock_irq()` is not, in fact, re-enabling IRQs. This seems like a miscompilation, unless there's something subtly wrong with `raw_spin_unlock_irq()`.

**Comment by osandov** [comment]
That "fix" depends on the wrappers introduced after v5.13, but this seems to work for v5.11-v5.13:

```diff
diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 4ca80df205ce6..4901f47e84d84 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -5150,7 +5150,8 @@ static void __sched notrace __schedule(bool preempt)
 
 		rq_unpin_lock(rq, &rf);
 		__balance_callbacks(rq);
-		raw_spin_unlock_irq(&rq->lock);
+		raw_spin_unlock(&rq->lock);
+		local_irq_enable();
 	}
 }
 
```

**Comment by brenns10** [comment]
WHAT

**Comment by brenns10** [comment]
Just because this is absolutely bonkers, let me try to explain this issue all the way.

A bug in GCC causes a missing `sti` instruction (enabling interrupts) in the `__schedule()` function, one of the most core parts of the kernel. Whatever code generation bug is responsible, only gets triggered by code between commits torvalds/linux@2558aacff8586699bcd248b406febb28b0a25de2 (merged in 5.11) and torvalds/linux@5cb9eaa3d274f75539077a28cf01e3563195fa53 (merged in 5.14). This compiler bug means that, when `__schedule()` is called, and `prev == next` (that is, the only task which is currently runnable is the current task), interrupts do not get re-enabled.

During the SMP boot code, as each CPU is brought online, it must compute its "bogomips" value, a rough estimate of the number of loop iterations per jiffy, used for short delays throughout the kernel. Most modern x86 processors have `X86_FEATURE_CONSTANT_TSC`, which indicates that all the CPUs have the same value as CPU 0, so the kernel skips this computation for each CPU normally. However, if the kernel is run as a guest, without KVM, then the basic x86 CPU which QEMU is emulating will not have this feature flag, forcing each CPU to compute a bogomips.

CPU 0 needs to wait on that bogomips calculation, so it enters a tight loop where it checks whether CPU 1 has successfully booted, and if not, it calls `schedule()`. Since there are no other tasks running, we hit the compiler bug: interrupts are disabled on CPU 0. But since this is such a tight loop, and `schedule()` has no other tasks to choose from, `schedule()` still returns, just with interrupts disabled. Sadly, this means that CPU 0 can no longer handle timer interrupts.

This is a problem for CPU 1, which needs to see `jiffies` get incremented as part of its bogomips computation. We're in a bit of a deadlock situation now: CPU 0 waits for CPU 1 to boot, and CPU 1 waits for CPU to handle the timer interrupt.

It seems that at some point, the code in schedule() must enable interrupts elsewhere, for a brief period of time, so the timer interrupts do get handled, but very rarely, and so jiffies rarely increments. This means that each CPU boots incredibly slowly, since it needs to wait for these rare moments where the jiffies increment.

If the CPUs all do manage to boot, over a very long period of time, then the boot process will proceed, and since this compiler bug was very specific, it seems that the other areas where the kernel re-enables interrupts work fine. So eventually, interrupts are re-enabled and CPU 0 doesn't have any lingering side-effects from this bug (since it's relatively uncommon for `prev == next` in `schedule()`).

This means that, for kernels 5.11-5.13, our CI ran for a very long time, but rarely failed... And this only on systems without KVM - on systems with KVM, all of this gets avoided due to `X86_FEATURE_CONSTANT_TSC`.

**Comment by osandov** [comment]
Ugh, my disassembly comparison was actually wrong. I was comparing the `.o` files, which don't show what functions they're calling. This is the actual comparison:

```diff
__schedule
--- bad_vmlinux
+++ good_vmlinux
@@ -240,7 +240,7 @@
 	call   <__switch_to_asm>
 	mov    %rax,%rdi
 	call   <finish_task_switch.isra.0>
-	jmp    <__schedule+0x43f>
+	jmp    <__schedule+0x440>
 	mov    0x880(%r12),%eax
 	and    $0xfffffffc,%eax
 	mov    %eax,0x880(%r12)
@@ -250,10 +250,11 @@
 	mov    %r12,%rdi
 	call   <__balance_callbacks>
 	mov    %r12,%rdi
-	call   <__raw_spin_unlock_irq>
+	call   <__raw_spin_unlock>
+	sti    
 	mov    -0x30(%rbp),%rax
 	sub    %gs:0x28,%rax
-	je     <__schedule+0x453>
+	je     <__schedule+0x454>
 	call   <__stack_chk_fail>
 	add    $0x28,%rsp
 	pop    %rbx
```

So there was a call to `__raw_spin_unlock_irq` before, which is:

```
ffffffff8104abd6 <__raw_spin_unlock_irq>:
ffffffff8104abd6:	c6 07 00             	movb   $0x0,(%rdi)
ffffffff8104abd9:	fb                   	sti    
ffffffff8104abda:	c3                   	ret    
```

Which _does_ have the `sti` instruction. (I disabled `CONFIG_PARAVIRT_SPINLOCKS` to rule that out, so the actual kernel in vmtest might look different.) So I'm back to not understanding what's going on.

Maybe we should just apply the workaround and move on with our lives...

**Comment by brenns10** [comment]
Yeah, I mean I think more time spent on this won't be helping us any more. It's just picking the workaround that's least impactful. I see three options:

1. Disable kernels 5.11 - 5.13
2. Apply your patch which mysteriously works, despite there not (??) being a compiler bug
3. Update `vmtest.vm` to set `lpj=10000000` when KVM is disabled. It's not accurate but not terribly awful.

I think your patch makes the most sense, as we can limit it to 5.11 - 5.13. Setting lpj would be quicker and easier (no need to rebuild the kernels) but impacts all the other kernels which aren't experiencing this.

I say let's ship your workaround patch and forget this bug ever existed.

**Comment by osandov** [comment]
I pushed my workaround patch and just kicked off a vmtest build.

**Comment by brenns10** [comment]
I triggered CI build on my fork after the new kernels were built and it worked with no big delay!
https://github.com/brenns10/drgn/actions/runs/6753434372

There is still a range, 17-30 minutes, which I find interesting, but clearly this issue is worked around successfully.

**Comment by osandov** [comment]
Great! It's still a mystery why this started suddenly, and what the root cause really is, but so be it. The vmtest build had one 5.19 VM that randomly died in the middle of booting, but I've seen that before. I hope GitHub Actions adds support for KVM someday.
